yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala

```diff
@@ -390,8 +393,22 @@ private[spark] class Client(
     // Upload Spark and the application JAR to the remote file system if necessary,
     // and add them as local resources to the application master.
     val fs = destDir.getFileSystem(hadoopConf)
-    val nns = YarnSparkHadoopUtil.get.getNameNodesToAccess(sparkConf) + destDir
-    YarnSparkHadoopUtil.get.obtainTokensForNamenodes(nns, hadoopConf, credentials)
+
+    // Merge credentials obtained from registered providers
+    var nearestTimeOfNextRenewal = Long.MaxValue
+    credentialManager.obtainCredentials(hadoopConf, credentials).foreach { t =>
```

Comment:
Yes, in the current implementation of `HDFSCredentialProvider`, time of next token renewal will only be calculated when principal is provided (`loginFromKeytab` is true).